File: src/android-common/src/main/java/com/geovault/common/sharing/ShareId.kt
package com.geovault.common.sharing

/**
 * UUID4 token used as the global share-link identity.
 */
data class ShareId(val value: String) {
    init {
        require(isShareId(value)) { "Share id must be a UUID4" }
    }

    override fun toString(): String = value

    companion object {
        private val SHARE_ID_PATTERN =
            Regex("^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$")

        fun parse(value: String?): ShareId? {
            val normalized = value?.trim()?.lowercase().orEmpty()
            if (!SHARE_ID_PATTERN.matches(normalized)) {
                return null
            }
            return ShareId(normalized)
        }

        fun isShareId(value: String?): Boolean = parse(value) != null
    }
}

File: src/backend/geo_lib/sharing/share_id.py
import re
import uuid
from collections.abc import Callable

SHARE_ID_PATTERN = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
)


class ShareId:
    """UUID4 token used as the global share-link identity."""

    def __init__(self, value: str):
        normalized = value.strip().lower()
        if not SHARE_ID_PATTERN.fullmatch(normalized):
            raise ValueError("Share id must be a UUID4")
        self.value = normalized

    def __str__(self) -> str:
        return self.value

    def __eq__(self, other: object) -> bool:
        if isinstance(other, ShareId):
            return self.value == other.value
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self.value)

    @classmethod
    def parse(cls, value: str | None) -> "ShareId | None":
        if not value or not isinstance(value, str):
            return None
        try:
            return cls(value)
        except ValueError:
            return None

    @classmethod
    def generate(cls) -> "ShareId":
        return cls(str(uuid.uuid4()))

    @classmethod
    def generate_unique(cls, exists: Callable[[str], bool]) -> "ShareId":
        candidate = cls.generate()
        while exists(candidate.value):
            candidate = cls.generate()
        return candidate
